### PR TITLE
device: stop using APPLICATION init level

### PR DESCRIPTION
**Pull request by gmarull**
This PR contains a series of patches that remove usage of `APPLICATION` init level on devices. Device drivers should be restricted to PRE_KERNEL_1/2 and POST_KERNEL init levels. It also does some other related cleanups, like using device model on some modules, or SYS_INIT in some others.

### CODE REVIEW

**Comment by carlescufi** [comment]
@gmarull should this target the collab-init branch instead?

**Review by carlescufi — DISMISSED** [review]

File: subsys/bluetooth/controller/coex/coex_ticker.c [hunk 1]
```diff
@@ -151,59 +140,42 @@ static void coex_ticker_grant_irq_handler(const struct device *dev,
...
-	.grant_spec = GPIO_DT_SPEC_GET(COEX_NODE, grant_gpios),
-	.grant_delay_us = DT_PROP(COEX_NODE, grant_delay_us)
-};
-static struct coex_ticker_data data;
-
-DEVICE_DEFINE(coex_ticker, "COEX_TICKER", &coex_ticker_init, NULL,
-	&data, &config,
-	APPLICATION, 90,
-	NULL);
+SYS_INIT(coex_ticker_init, APPLICATION, 90);
```
Comment (fabiobaltieri) [inline subsys/bluetooth/controller/coex/coex_ticker.c]: Hey, I think this should rather become a `DEVICE_DT_DEFINE`: the driver may not make much use of the device structure, but it is based on a devicetree structure, if you instantiate it with `DEVICE_DT_DEFINE` then the corresponding struct device would get an ordinal, and the buildsystem will ensure that the init function get called after the corresponding device (the gpio controller in `grant-gpios`) have been already initialized.

Comment (gmarull) [reply subsys/bluetooth/controller/coex/coex_ticker.c]: hmm good point, let me check how it's organized in dt

**Comment by gmarull** [comment]
I wonder if we should forbid EARLY/SMP (first couldn't be used anyway), and deprecate APPLICATION, considering the number of cases I've found.

File: subsys/bluetooth/controller/coex/Kconfig [hunk 2]
```diff
@@ -5,8 +5,7 @@
 
 menuconfig BT_CTLR_COEX_DRIVERS
 	bool "Bluetooth Co-existence Drivers"
-	default n
-	depends on BT_CTLR
+	depends on BT_CTLR && DT_HAS_GPIO_RADIO_COEX_ENABLED
```
Comment (fabiobaltieri) [inline subsys/bluetooth/controller/coex/Kconfig:8]: `default y` and drop `CONFIG_BT_CTLR_COEX_DRIVERS=y` from `samples/bluetooth/beacon/prj-coex.conf`?

File: include/zephyr/device.h [hunk 3]
```diff
@@ -942,6 +948,9 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
  * @param prio Initialization priority.
  */
 #define Z_DEVICE_INIT_ENTRY_DEFINE(node_id, dev_id, init_fn_, level, prio)     \
+	BUILD_ASSERT(IS_ENABLED(Z_DEVICE_LEVEL_##level),                       \
+		     "Invalid level: " STRINGIFY(level));                      \
```
Comment (fabiobaltieri) [inline include/zephyr/device.h:952]: Maybe "Unsupported level for device drivers:"? I feel like "invalid level" suggests that like the name is wrong or something else.

**Comment by henrikbrixandersen** [comment]
> > @gmarull should this target the collab-init branch instead?
> 
> no, I consider this prep-work

I really think this ought to go into the `colab-init` branch instead and be rolled into `main` at the start of the v3.6 development cycle. While this is indeed a clean-up or prep work, the idea of the `colab-init` branch was to avoid breaking downstream users across a release. Imposing limits on the currently used init levels can break downstream code.

**Comment by gmarull** [comment]
> > > @gmarull should this target the collab-init branch instead?
> > 
> > 
> > no, I consider this prep-work
> 
> I really think this ought to go into the `colab-init` branch instead and be rolled into `main` at the start of the v3.6 development cycle. While this is indeed a clean-up or prep work, the idea of the `colab-init` branch was to avoid breaking downstream users across a release. Imposing limits on the currently used init levels can break downstream code.

Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.

File: subsys/bluetooth/controller/coex/Kconfig (hunk 2 above)
Comment (gmarull) [reply subsys/bluetooth/controller/coex/Kconfig:9]: done

File: include/zephyr/device.h (hunk 3 above)
Comment (gmarull) [reply include/zephyr/device.h]: done

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by dcpleung** [comment]
> Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.

FYI, SMP init level is used by the `ipm_cavs_idc.c` driver. The `CONFIG_SMP_BOOT_DELAY` guard is probably incorrect but the driver uses SMP init level.

**Comment by gmarull** [comment]
> > Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.
> 
> ~FYI, SMP init level is used by the `ipm_cavs_idc.c` driver. The `CONFIG_SMP_BOOT_DELAY` guard is probably incorrect but the driver uses SMP init level.~ Scratch that... the driver is not being used at all, even on SOF. I will make a PR to remove it.

Note that this PR only applies to devices, not sys_inits.

**Comment by nashif** [comment]
> 
> Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.

```
EARLY
  __init_soc_num_cpus_init: soc_num_cpus_init(NULL)
  __init___device_dts_ord_73: mem_win_init(__device_dts_ord_73)
  __init___device_dts_ord_74: mem_win_init(__device_dts_ord_74)
  __init___device_dts_ord_75: mem_win_init(__device_dts_ord_75)
  __init___device_dts_ord_76: mem_win_init(__device_dts_ord_76)
  __init_winstream_console_init: winstream_console_init(NULL)
  __init_boot_complete: boot_complete(NULL)
PRE_KERNEL_1
  __init___device_dts_ord_43: dw_ace_init(__device_dts_ord_43)
  __init_soc_init: soc_init(NULL)
PRE_KERNEL_2
  __init___device_dts_ord_44: dt_init(__device_dts_ord_44)
  __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
POST_KERNEL
  __init___device_dts_ord_90: sys_mm_drv_mm_init(__device_dts_ord_90)
  __init_malloc_prepare: malloc_prepare(NULL)
APPLICATION
SMP
```

**Comment by gmarull** [comment]
> > Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.
> 
> ```
> EARLY
>   __init_soc_num_cpus_init: soc_num_cpus_init(NULL)
>   __init___device_dts_ord_73: mem_win_init(__device_dts_ord_73)
>   __init___device_dts_ord_74: mem_win_init(__device_dts_ord_74)
>   __init___device_dts_ord_75: mem_win_init(__device_dts_ord_75)
>   __init___device_dts_ord_76: mem_win_init(__device_dts_ord_76)
>   __init_winstream_console_init: winstream_console_init(NULL)
>   __init_boot_complete: boot_complete(NULL)
> PRE_KERNEL_1
>   __init___device_dts_ord_43: dw_ace_init(__device_dts_ord_43)
>   __init_soc_init: soc_init(NULL)
> PRE_KERNEL_2
>   __init___device_dts_ord_44: dt_init(__device_dts_ord_44)
>   __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
> POST_KERNEL
>   __init___device_dts_ord_90: sys_mm_drv_mm_init(__device_dts_ord_90)
>   __init_malloc_prepare: malloc_prepare(NULL)
> APPLICATION
> SMP
> ```

Err, EARLY can’t be used for devices as stated in the PR. Will look into this…

**Comment by gmarull** [comment]
Note: I have dropped 6f6a45e, so this PR is now just a cleanup of APPLICATION level usage. Will re-take the deprecation work after https://github.com/zephyrproject-rtos/zephyr/issues/62627 is fixed.

**Review by fabiobaltieri — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Comment by gmarull** [comment]
last push: fixed some issues detected by the check_init_priorities script

**Review by fabiobaltieri — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by nashif — APPROVED** [review]
this looks good, but please lets get more reviewers before we merge this and probably put a bit more details about the big picture of why this is being done and where we are going with this. 

As agreed in the Architecture Review meeting, we wanted to have a full plan with details of the init level rework (read: RFC), something we can reference in many of those PRs.

**Review by jhedberg — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]
display, usb LGTM

**Review by henrikbrixandersen — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by nashif** [comment]
@gmarull  how are we going to prevent this from happening going forward and deprecate the usage? i.e. if someone tries to do that, should we emit a warning somewhere?

**Comment by gmarull** [comment]
> @gmarull how are we going to prevent this from happening going forward and deprecate the usage? i.e. if someone tries to do that, should we emit a warning somewhere?

Will open a PR deprecating this init level for devices (was initially part of this PR, but wanted to skip controversies for now)

**Pull request merged**